internal/rpc/otel_interceptor.go

```diff
@@ -0,0 +1,420 @@
+	otelFtlVerbRef          = "ftl.verb.ref"
+	otelFtlVerbModule       = "ftl.verb.module"
+	otelMessageEvent        = "message"
+	otelMessageIDKey        = attribute.Key("message.id")
+	otelMessageSizeKey      = attribute.Key("message.uncompressed_size")
+	otelMessageTypeKey      = attribute.Key("message.type")
+	otelMessageTypeSent     = "SENT"
+	otelMessageTypeReceived = "RECEIVED"
+)
+
+func OtelInterceptor() connect.Interceptor {
+	return &otelInterceptor{}
+}
+
+type otelInterceptor struct{}
+
+var clientInstruments instrumentation
+var serverInstruments instrumentation
+
+func init() {
+	clientInstruments = createInstruments(otel.GetMeterProvider().Meter("ftl.rpc.client"))
+	serverInstruments = createInstruments(otel.GetMeterProvider().Meter("ftl.rpc.server"))
+}
+
+func getInstruments(isClient bool) instrumentation {
+	if isClient {
+		return clientInstruments
+	}
+	return serverInstruments
+}
+
+func getAttributes(ctx context.Context, rpcSystemKey string) []attribute.KeyValue {
+	logger := log.FromContext(ctx)
+	attributes := []attribute.KeyValue{
+		semconv.RPCSystemKey.String(rpcSystemKey),
+	}
+	requestKey, err := RequestKeyFromContext(ctx)
+	if err != nil {
+		logger.Debugf("failed to get request key: %s", err)
+	}
+	if key, ok := requestKey.Get(); ok {
+		attributes = append(attributes, attribute.String(otelFtlRequestKey, key.String()))
+	}
+	if verb, ok := VerbFromContext(ctx); ok {
+		attributes = append(attributes, attribute.String(otelFtlVerbRef, verb.String()))
+		attributes = append(attributes, attribute.String(otelFtlVerbModule, verb.Module))
+	}
+	return attributes
+}
+
+func (i *otelInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
+	return func(ctx context.Context, request connect.AnyRequest) (connect.AnyResponse, error) {
+		ctx = propagateOtelHeaders(ctx, request.Spec().IsClient, request.Header())
+		requestStartTime := time.Now()
+		isClient := request.Spec().IsClient
+		name := strings.TrimLeft(request.Spec().Procedure, "/")
+
+		spanKind := trace.SpanKindClient
+		requestSpan := otelMessageTypeKey.String(otelMessageTypeSent)
+		responseSpan := otelMessageTypeKey.String(otelMessageTypeReceived)
+		if !isClient {
+			spanKind = trace.SpanKindServer
+			requestSpan = otelMessageTypeKey.String(otelMessageTypeReceived)
+			responseSpan = otelMessageTypeKey.String(otelMessageTypeSent)
+		}
+
+		attributes := getAttributes(ctx, request.Peer().Protocol)
+		traceOpts := []trace.SpanStartOption{
+			trace.WithAttributes(attributes...),
+			trace.WithSpanKind(spanKind),
+		}
+		tracer := otel.GetTracerProvider().Tracer(request.Spec().Procedure)
+		ctx, span := tracer.Start(ctx, name, traceOpts...)
+		defer span.End()
+
+		var requestSize int
+		if request != nil {
+			if msg, ok := request.Any().(proto.Message); ok {
+				requestSize = proto.Size(msg)
+			}
+		}
+
+		span.AddEvent(otelMessageEvent,
+			trace.WithAttributes(
+				requestSpan,
+				otelMessageIDKey.Int(1),
```

Comment:
What's message ID for?

Reply:
Doing this for consistency across the rpc interceptors. The individual messages are all shown in the span event, and their ID's must be unique, in the unary case there is one request and one response